File: java-automation/src/test/java/com/myecommerce/automation/driver/api/MyEcommerceDriver.java
package com.myecommerce.automation.driver.api;

import com.fasterxml.jackson.databind.DeserializationFeature;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.PropertyNamingStrategies;
import com.myecommerce.automation.dsl.domain.DeliveryOption;
import com.myecommerce.automation.dsl.domain.DeliveryState;
import com.myecommerce.automation.dsl.domain.ProductCard;
import com.myecommerce.automation.dsl.domain.ProductDetail;
import com.myecommerce.automation.dsl.domain.ProductListing;
import com.myecommerce.automation.dsl.domain.SearchResults;
import com.myecommerce.automation.dsl.protocols.CatalogueProtocol;
import com.myecommerce.automation.dsl.protocols.Channel;
import com.myecommerce.automation.dsl.protocols.DriverRegistry;
import io.restassured.RestAssured;
import io.restassured.common.mapper.TypeRef;
import io.restassured.config.ObjectMapperConfig;
import lombok.extern.java.Log;
import net.serenitybdd.rest.SerenityRest;

import java.util.List;
import java.util.stream.IntStream;

@Log
public final class MyEcommerceDriver implements CatalogueProtocol {

    static {
        DriverRegistry.register(Channel.API, MyEcommerceDriver::new);
        RestAssured.config = RestAssured.config().objectMapperConfig(
            ObjectMapperConfig.objectMapperConfig()
                .jackson2ObjectMapperFactory((cls, charset) ->
                    new ObjectMapper()
                        .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
                        .setPropertyNamingStrategy(PropertyNamingStrategies.SNAKE_CASE)));
    }

    private static final String BASE_URL = "http://localhost:8001";

    private int currentProductId = -1;
    private String lastSearchTerm = "";

    @Override public void browseCatalogue() {}
    @Override public void returnToProductListing() {}
    @Override public void chooseAlternativeDeliveryOption() {}

    @Override
    public void viewFirstProduct() {
        this.currentProductId = fetchProducts("").getFirst().id();
        log.info("selected product id=%d".formatted(currentProductId));
    }

    @Override
    public void searchFor(String term) {
        this.lastSearchTerm = term;
    }

    @Override
    public ProductListing getProductListing() {
        var cards = fetchProducts("").stream()
            .map(product -> new ProductCard(product.title(), formatPrice(product.price()), absoluteImageUrl(product.imageUrl())))
            .toList();
        log.info("%d products".formatted(cards.size()));
        return new ProductListing(cards, false);
    }

    @Override
    public ProductDetail getProductDetail() {
        var product = fetchProductDetail(currentProductId);
        log.info("title='%s'".formatted(product.title()));
        return ProductDetail.builder()
            .title(product.title())
            .price(formatPrice(product.price()))
            .description(product.description())
            .imagePresent(product.imageUrl() != null && !product.imageUrl().isBlank())
            .addToCartButtonText("Add to Cart")
            .addToCartEnabled(true)
            .build();
    }

    @Override
    public DeliveryState getDeliveryState() {
        var activeDeliveryOptions = fetchActiveDeliveryOptions();
        if (activeDeliveryOptions.isEmpty()) {
            log.info("no active delivery options");
            return DeliveryState.builder()
                .sectionVisible(false)
                .options(List.of())
                .headerText("")
                .minimumOrderTextPresent(false)
                .build();
        }
        log.info("%d active delivery options".formatted(activeDeliveryOptions.size()));
        return DeliveryState.builder()
            .sectionVisible(true)
            .options(toDeliveryOptions(activeDeliveryOptions))
            .headerText("Delivery Options")
            .minimumOrderTextPresent(false)
            .build();
    }

    @Override
    public SearchResults getSearchResults() {
        var cards = fetchProducts(lastSearchTerm).stream()
            .map(product -> new ProductCard(product.title(), formatPrice(product.price()), absoluteImageUrl(product.imageUrl())))
            .toList();
        log.info("%d results for '%s'".formatted(cards.size(), lastSearchTerm));
        return new SearchResults(cards, cards.isEmpty());
    }

    @Override
    public String currentUrl() {
        if (!lastSearchTerm.isBlank()) return BASE_URL + "/products?search=" + lastSearchTerm;
        if (currentProductId >= 0)     return BASE_URL + "/products/" + currentProductId;
        return BASE_URL + "/products";
    }

    private List<ApiDeliveryOption> fetchActiveDeliveryOptions() {
        return fetchProductDetail(currentProductId).deliveryOptions().stream()
            .filter(ApiDeliveryOption::isActive)
            .toList();
    }

    private List<DeliveryOption> toDeliveryOptions(List<ApiDeliveryOption> activeOptions) {
        return IntStream.range(0, activeOptions.size())
            .mapToObj(index -> new DeliveryOption(activeOptions.get(index).name(), index == 0))
            .toList();
    }

    private List<ApiProduct> fetchProducts(String searchTerm) {
        var spec = SerenityRest.given().baseUri(BASE_URL);
        if (!searchTerm.isBlank()) {
            spec = spec.queryParam("search", searchTerm);
        }
        return spec.get("/products")
            .then().statusCode(200)
            .extract().as(new TypeRef<>() {});
    }

    private ApiProductDetail fetchProductDetail(int id) {
        return SerenityRest.given().baseUri(BASE_URL)
            .get("/products/" + id)
            .then().statusCode(200)
            .extract().as(ApiProductDetail.class);
    }

    private String formatPrice(double price) {
        return "$%.2f".formatted(price);
    }

    private String absoluteImageUrl(String imageUrl) {
        if (imageUrl == null || imageUrl.isBlank()) return "";
        if (imageUrl.startsWith("http")) return imageUrl;
        return BASE_URL + imageUrl;
    }
}

File: python-automation/src/automation/driver/api/my_ecommerce_driver.py
from automation.driver.api.http_holder import get_http
from automation.driver.ports.http_port import HttpPort
from automation.dsl.domain import (
    DeliveryOption,
    DeliveryState,
    ProductCard,
    ProductDetail,
    ProductListing,
    SearchResults,
)
from automation.dsl.protocols import driver_registry
from automation.dsl.protocols.catalogue_protocol import CatalogueProtocol
from automation.dsl.protocols.channel import Channel

BASE_URL = "http://localhost:8001"


class ApiMyEcommerceDriver(CatalogueProtocol):

    def __init__(self, http: HttpPort) -> None:
        self._http = http
        self._current_product_id = -1
        self._last_search_term = ""

    def browse_catalogue(self) -> None:
        return None

    def return_to_product_listing(self) -> None:
        return None

    def choose_alternative_delivery_option(self) -> None:
        return None

    def view_first_product(self) -> None:
        products = self._fetch_products("")
        if not products:
            raise RuntimeError("No products available to view")
        self._current_product_id = int(products[0]["id"])

    def search_for(self, term: str) -> None:
        self._last_search_term = term

    def get_product_listing(self) -> ProductListing:
        cards = [self._to_card(p) for p in self._fetch_products("")]
        return ProductListing(cards=cards, has_visible_loading_indicators=False)

    def get_product_detail(self) -> ProductDetail:
        product = self._fetch_product_detail(self._current_product_id)
        image_url = product.get("imageUrl") or ""
        return ProductDetail(
            title=product.get("title", ""),
            price=self._format_price(product.get("price", 0)),
            description=product.get("description", ""),
            image_present=bool(image_url.strip()),
            add_to_cart_button_text="Add to Cart",
            add_to_cart_enabled=True,
        )

    def get_delivery_state(self) -> DeliveryState:
        active = self._fetch_active_delivery_options()
        if not active:
            return DeliveryState(
                section_visible=False,
                options=[],
                header_text="",
                minimum_order_text_present=False,
            )
        options = [
            DeliveryOption(label=opt.get("name", ""), selected=(i == 0))
            for i, opt in enumerate(active)
        ]
        return DeliveryState(
            section_visible=True,
            options=options,
            header_text="Delivery Options",
            minimum_order_text_present=False,
        )

    def get_search_results(self) -> SearchResults:
        cards = [self._to_card(p) for p in self._fetch_products(self._last_search_term)]
        return SearchResults(cards=cards, empty_state_visible=not cards)

    def current_url(self) -> str:
        if self._last_search_term.strip():
            return f"{BASE_URL}/products?search={self._last_search_term}"
        if self._current_product_id >= 0:
            return f"{BASE_URL}/products/{self._current_product_id}"
        return f"{BASE_URL}/products"

    def _fetch_products(self, search_term: str) -> list[dict]:
        if not search_term.strip():
            return self._http.get_list_as("/products")
        return self._http.get_list_with_query_as("/products", "search", search_term)

    def _fetch_product_detail(self, product_id: int) -> dict:
        return self._http.get_as(f"/products/{product_id}")

    def _fetch_active_delivery_options(self) -> list[dict]:
        product = self._fetch_product_detail(self._current_product_id)
        delivery_options = product.get("deliveryOptions") or []
        return [opt for opt in delivery_options if opt.get("isActive")]

    def _to_card(self, p: dict) -> ProductCard:
        return ProductCard(
            title=p.get("title", ""),
            price=self._format_price(p.get("price", 0)),
            image_url=self._absolute_image_url(p.get("imageUrl") or ""),
        )

    @staticmethod
    def _format_price(price: float | int) -> str:
        return f"${price:.2f}"

    @staticmethod
    def _absolute_image_url(image_url: str) -> str:
        if not image_url.strip():
            return ""
        if image_url.startswith("http"):
            return image_url
        return f"{BASE_URL}{image_url}"


def _build_driver() -> ApiMyEcommerceDriver:
    return ApiMyEcommerceDriver(get_http())


driver_registry.register(Channel.API, _build_driver)
